File: src/ros2_cpp_tutorial/src/p2_fast_agent.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
using std::placeholders::_1;

using namespace std::chrono_literals;

/* This example creates a subclass of Node and uses std::bind() to register a
* member function as a callback from the timer. */

class FastAgent : public rclcpp::Node
{
  public:
    FastAgent()
    : Node("fast_agent"), count_(0)
    {
      publisher_ = this->create_publisher<std_msgs::msg::String>("fast_topic", 10);
      timer_ = this->create_wall_timer(
        500ms, std::bind(&FastAgent::timer_callback, this));
      subscription_ = this->create_subscription<std_msgs::msg::String>(
        "slow_topic", 10, std::bind(&FastAgent::topic_callback, this, _1));
    }

  private:
    void timer_callback()
    {
      auto message = std_msgs::msg::String();
      message.data = "Fast timer " + std::to_string(count_++);
      RCLCPP_INFO(this->get_logger(), "[Fast] Publishing: '%s'", message.data.c_str());
      publisher_->publish(message);
    }

    void topic_callback(const std_msgs::msg::String & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "[Fast] I heard: '%s'", msg.data.c_str());
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
    size_t count_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FastAgent>());
  rclcpp::shutdown();
  return 0;
}

File: src/ros2_python_tutorial/ros2_python_tutorial/p2_fast_agent.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class FastAgent(Node):

    def __init__(self):
        super().__init__('fast_agent')
        self.publisher_ = self.create_publisher(String, 'fast_topic', 10)
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.subscription = self.create_subscription(
            String,
            'slow_topic',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning
        self.i = 0

    def timer_callback(self):
        msg = String()
        msg.data = 'Fast timer %d' % self.i
        self.publisher_.publish(msg)
        self.get_logger().info('[Fast] Publishing: "%s"' % msg.data)
        self.i += 1

    def listener_callback(self, msg):
        self.get_logger().info('[Fast] I heard: "%s"' % msg.data)


def main(args=None):
    rclpy.init(args=args)

    fast_agent = FastAgent()

    rclpy.spin(fast_agent)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    fast_agent.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()